### ISSUE TITLE
If minibatchSize is different when training and testing the SameDiffOutputLayer, it will throw an exception, because labels have a different 0 dimension

### ISSUE THREAD

**Issue by longzhendong** [eclipse/deeplearning4j#7844]
* 1.0.0-beta4
* windows 10

If minibatchSize is different when training and testing the SameDiffOutputLayer, it will throw an exception, because labels have a different 0 dimension

Exception information is as follows：
MULTIPLY OP: the shapes of x [10, 10] and y [20, 10] are not suitable for broadcast !
Exception in thread "main" java.lang.RuntimeException: Op [multiply] execution failed
	at org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner.exec(NativeOpExecutioner.java:1645)
	at org.nd4j.autodiff.samediff.internal.InferenceSession.getOutputs(InferenceSession.java:449)
	at org.nd4j.autodiff.samediff.internal.InferenceSession.getOutputs(InferenceSession.java:37)
	at org.nd4j.autodiff.samediff.internal.AbstractSession.output(AbstractSession.java:250)
	at org.nd4j.autodiff.samediff.SameDiff.exec(SameDiff.java:3932)
	at org.nd4j.autodiff.samediff.SameDiff.execSingle(SameDiff.java:3901)
	at org.nd4j.autodiff.samediff.SameDiff.execAndEndResult(SameDiff.java:3193)
	at org.deeplearning4j.nn.layers.samediff.SameDiffOutputLayer.activateHelper(SameDiffOutputLayer.java:110)
	at org.deeplearning4j.nn.layers.samediff.SameDiffOutputLayer.activate(SameDiffOutputLayer.java:84)
	at org.deeplearning4j.nn.layers.AbstractLayer.activate(AbstractLayer.java:257)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.outputOfLayerDetached(MultiLayerNetwork.java:1273)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.doEvaluationHelper(MultiLayerNetwork.java:3348)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.doEvaluation(MultiLayerNetwork.java:3300)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.evaluate(MultiLayerNetwork.java:3493)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.evaluate(MultiLayerNetwork.java:3403)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.evaluate(MultiLayerNetwork.java:3235)
	at org.deeplearning4j.examples.samediff.dl4j.Ex1BasicSameDiffOutputLayerExample.main(Ex1BasicSameDiffOutputLayerExample.java:102)
Caused by: java.lang.RuntimeException: Op validation failed
	at org.nd4j.nativeblas.Nd4jCpu$NativeOps.execCustomOp(Native Method)
	at org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner.exec(NativeOpExecutioner.java:2045)
	at org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner.exec(NativeOpExecutioner.java:1637)
	... 16 more

**Comment by longzhendong** [comment]
@AlexDBlack

**Comment by AlexDBlack** [comment]
I've been unable to reproduce this based on your description. Can you share some code to reproduce the problem?

**Comment by longzhendong** [comment]
@AlexDBlack 
public class MinimalSameDiffOutputLayer extends SameDiffOutputLayer {

	private static final long serialVersionUID = 1L;

	private WeightInit weightInit;

	private int nIn;
	private int nOut;
	private Activation activation;

	public MinimalSameDiffOutputLayer(int nIn, int nOut, Activation activation, WeightInit weightInit) {
		this.nIn = nIn;
		this.nOut = nOut;
		this.activation = activation;
		this.weightInit = weightInit;
	}

	@Override
	public SDVariable defineLayer(SameDiff sameDiff, SDVariable layerInput, SDVariable labels,
			Map<String, SDVariable> paramTable) {
		SDVariable weights = paramTable.get(DefaultParamInitializer.WEIGHT_KEY);
		SDVariable bias = paramTable.get(DefaultParamInitializer.BIAS_KEY);
		SDVariable mmul = sameDiff.mmul("mmul", layerInput, weights);
		SDVariable z = mmul.add("z", bias);
		SDVariable a = activation.asSameDiff("out", sameDiff, z);
		// 交叉熵记得前面的负号
		return labels.mul(sameDiff.math().log(a)).neg().sum(1);

	}

	@Override
	public String activationsVertexName() {
		return "out";
	}

	@Override
	public void defineParameters(SDLayerParams params) {
		params.addWeightParam(DefaultParamInitializer.WEIGHT_KEY, nIn, nOut);
		params.addBiasParam(DefaultParamInitializer.BIAS_KEY, 1, nOut);
	}

	@Override
	public void initializeParameters(Map<String, INDArray> params) {
		params.get(DefaultParamInitializer.BIAS_KEY).assign(0);
		initWeights(nIn, nOut, weightInit, params.get(DefaultParamInitializer.WEIGHT_KEY));

	}

	@Override
	public InputType getOutputType(int layerIndex, InputType inputType) {
		return InputType.feedForward(nOut);
	}

	public int getNIn() {
		return nIn;
	}

	public void setNIn(int nIn) {
		this.nIn = nIn;
	}

	public int getNOut() {
		return nOut;
	}

	public void setNOut(int nOut) {
		this.nOut = nOut;
	}

	public Activation getActivation() {
		return activation;
	}

	public void setActivation(Activation activation) {
		this.activation = activation;
	}

}

public class Ex1BasicSameDiffOutputLayerExample {

    public static void main(String[] args) throws Exception {

        int networkNumInputs = 28*28;       //For MNIST - 28x28 pixels
        int networkNumOutputs = 10;         //For MNIST - 10 classes
        int layerSize = 50;                //128 units for the SameDiff layers
        
        
        MinimalSameDiffDense  dense=    new MinimalSameDiffDense(layerSize, layerSize, Activation.SIGMOID, WeightInit.XAVIER);
       // dense.setIDropout(new Dropout(0.9));

        MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder().dataType(DataType.DOUBLE)
            .updater(new RmsProp(0.01))
            .seed(12345)
            .list()
            //Add two custom layers:
            .layer(new MinimalSameDiffDense(networkNumInputs, layerSize, Activation.SIGMOID, WeightInit.XAVIER))
            .layer(dense)
            //Combine with a standard DL4J output layer
            .layer(new MinimalSameDiffOutputLayer(layerSize,networkNumOutputs,Activation.SOFTMAX,WeightInit.XAVIER))
            .build();

        MultiLayerNetwork net = new MultiLayerNetwork(conf);
        net.init();
        net.output(Nd4j.rand(DataType.DOUBLE, 1, 784));	
      
     
        System.out.println(net.summary());
        
        UIServer uiServer = UIServer.getInstance();
        StatsStorage statsStorage = new InMemoryStatsStorage();
        uiServer.attach(statsStorage);
        net.setListeners(new ScoreIterationListener(50), new StatsListener( statsStorage,20));
        //Train and evaluate the network with the custom SameDiff layer
        //Note that training and evaluation is the same as with built-in layers
        DataSetIterator train = new MnistDataSetIterator(10, true, 12345);
        net.fit(train, 1);  //Train for 1 epoch

        DataSetIterator test = new MnistDataSetIterator(20, false, 12345);
        net.clear();
        Evaluation e = net.evaluate(test);
        System.out.println(e.stats());

        //Also: validate correctness of the network/layer
       // validateLayer();
    }

**Comment by AlexDBlack** [comment]
Thanks for reporting.
This is finally fixed here, and will be merged into eclipse/deeplearning4j master soon: https://github.com/SkymindIO/deeplearning4j/pull/28
